components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java

```diff
@@ -754,85 +761,101 @@ private void updateConsentPurposeConfiguration(Connection connection, int applic
-                        certificateReferenceId = serviceProviderProperties[i].getValue();
-                        break;
-                    }
-                }
+        if (spProperties == null) {
+            return;
+        }
 
-                // If there is a certificate reference, remove it from the properties array.
-                // Removing will be done by creating a new array and copying the elements other than the
-                // certificate reference from the existing array,
-                if (certificateReferenceIdIndex > -1) {
+        OptionalInt certificateReferenceIdIndex = IntStream.range(0, spProperties.length)
+                .filter(index -> SP_PROPERTY_NAME_CERTIFICATE.equals(spProperties[index].getName()))
+                .findFirst();
 
-                    ServiceProviderProperty[] propertiesWithoutCertificateReference =
-                            new ServiceProviderProperty[serviceProviderProperties.length - 1];
+        if (certificateReferenceIdIndex.isPresent()) {
+            int certificateReferenceIndex = certificateReferenceIdIndex.getAsInt();
+            int certificateID = Integer.parseInt(spProperties[certificateReferenceIdIndex.getAsInt()].getValue());
 
-                    System.arraycopy(serviceProviderProperties, 0, propertiesWithoutCertificateReference,
-                            0, certificateReferenceIdIndex);
-                    System.arraycopy(serviceProviderProperties, certificateReferenceIdIndex + 1,
-                            propertiesWithoutCertificateReference, certificateReferenceIdIndex,
-                            propertiesWithoutCertificateReference.length - certificateReferenceIdIndex);
+            serviceProvider.setSpProperties(getFilteredSpProperties(spProperties, certificateReferenceIndex));
+            deleteCertificate(certificateID, IdentityTenantUtil.getTenantDomain(tenantID));
+        }
+    }
 
-                    serviceProvider.setSpProperties(propertiesWithoutCertificateReference);
-                    deleteCertificate(connection, Integer.parseInt(certificateReferenceId));
-                }
-            }
-        } else {
-            // First get the certificate reference from the application properties.
-            ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+    /**
+     * Returns a new array of service provider properties with the property at the specified index removed.
+     * This method creates a copy of the original `spProperties` array, omitting the element at `indexToRemove`.
+     *
+     * @param spProperties   An array of service provider properties.
+     * @param indexToRemove  The index of the property to be removed from the array.
+     * @return A new array of service provider properties, excluding the property at the specified index.
+     */
+    private static ServiceProviderProperty[] getFilteredSpProperties(ServiceProviderProperty[] spProperties,
```

Comment:
why this needs to be a static ? I don't see a reason for that.

Reply:
Fixed with https://github.com/wso2/carbon-identity-framework/pull/6087/commits/9a2dc889d25ec1c2d6d950643a6310f1717fd44a